internal/servers/repository_workerauth.go

```diff
@@ -0,0 +1,637 @@
+package servers
+
+import (
+	"context"
+	"crypto/x509"
+	"database/sql"
+	"fmt"
+	"strconv"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/db/timestamp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/servers/store"
+	"github.com/hashicorp/boundary/internal/types/scope"
+	"github.com/hashicorp/go-dbw"
+	wrapping "github.com/hashicorp/go-kms-wrapping/v2"
+	nodee "github.com/hashicorp/nodeenrollment"
+	"github.com/hashicorp/nodeenrollment/types"
+	"google.golang.org/protobuf/proto"
+)
+
+// Ensure we implement the Storage interfaces
+var (
+	_ nodee.Storage              = (*WorkerAuthRepositoryStorage)(nil)
+	_ nodee.TransactionalStorage = (*WorkerAuthRepositoryStorage)(nil)
+)
+
+// WorkerAuthRepositoryStorage is the Worker Auth database repository
+type WorkerAuthRepositoryStorage struct {
+	reader db.Reader
+	writer db.Writer
+	kms    *kms.Kms
+	worker *store.Worker
+}
+
+// NewRepository creates a new Worker Auth WorkerAuthRepositoryStorage that implements the Storage interface
+// If used to authenticate and store a workerAuth record, a worker must be passed
+func NewRepositoryStorage(r db.Reader, w db.Writer, kms *kms.Kms, wrk *store.Worker) (*WorkerAuthRepositoryStorage, error) {
+	const op = "workerauth.NewRepository"
+	if r == nil {
+		return nil, errors.NewDeprecated(errors.InvalidParameter, op, "nil reader")
+	}
+	if w == nil {
+		return nil, errors.NewDeprecated(errors.InvalidParameter, op, "nil writer")
+	}
+	if kms == nil {
+		return nil, errors.NewDeprecated(errors.InvalidParameter, op, "nil kms")
+	}
+
+	return &WorkerAuthRepositoryStorage{
+		reader: r,
+		writer: w,
+		kms:    kms,
+		worker: wrk,
+	}, nil
+}
+
+// Flush is called when storage is done being performed. The boolean parameter
+// indicates whether the operation was successful (true) or failed (false).
+func (r *WorkerAuthRepositoryStorage) Flush(b bool) error {
```

Comment:
This function needs to gate the transaction for the storage operation(s). Is this going to be built out in a future PR?